lib/internal/shutil.js

```diff
@@ -0,0 +1,294 @@
+const fs = require('fs');
+const { join } = require('path');
+const { validatePath } = require('internal/fs/utils');
+const {
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+    ERR_INVALID_CALLBACK
+  }
+} = require('internal/errors');
+
+const isWindows = process.platform === 'win32';
+const _0666 = parseInt('666', 8);
+
+// for EMFILE handling
+let timeout = 0;
+
+function rmtree(path, options, callback) {
+  if (typeof options === 'function') {
+    callback = options;
+    options = {};
+  }
+
+  validatePath(path);
+  validateCallback(callback);
+
+  options = getOptions(options);
+
+  let busyTries = 0;
+
+  rmtree_(path, options, function CB(er) {
+    if (er) {
+      if (
+        (er.code === 'EBUSY' || er.code === 'ENOTEMPTY' || er.code === 'EPERM') &&
+        busyTries < options.maxBusyTries
+      ) {
+        busyTries++;
+        let time = busyTries * 100;
+        // try again, with the same exact callback as this one.
+        return setTimeout(function() {
+          rmtree_(path, options, CB);
+        }, time);
+      }
+
+      // this one won't happen if graceful-fs is used.
+      if (er.code === 'EMFILE' && timeout < options.emfileWait) {
+        return setTimeout(function() {
+          rmtree_(path, options, CB);
+        }, timeout++);
+      }
+
+      // already gone
+      if (er.code === 'ENOENT') er = null;
+
+      callback(er);
+    }
+
+    timeout = 0;
+    callback();
+  });
+}
+
+// Two possible strategies.
+// 1. Assume it's a file.  unlink it, then do the dir stuff on EPERM or EISDIR
```

Comment:
Curious why we're calling this `shutil` and not `rmraf`?

Reply:
I think that suggestion originally came from @refack. Python puts `rmtree` and number of other shell type functions like `copytree` under a lib called `shutil`. He suggested we do the same as eventually we may want to add more functionality to this lib.

Reply:
chiming in: yes, eventually some major portion of python's `shutil` should be implemented.  we want to start with this one

Reply:
Even if the majority of `shutil` ends up being ported, this belongs in `fs`.  Most of `shutil`'s functionality is already implemented in the `fs` module.  (Copyfile, readfile, mkdir, mkdir -p, etc.)  It adds unnecessary friction for Node.js users to have their fs utils split across two modules.

Reply:
In other words, from node's point of view, "shell utilities" are not a thing.  There are file system utilities in fs, and process utilities in child_process.

Reply:
hoping @refack can chime in on the reasoning

Reply:
I followed the original idea of having a `shutil` package inspired on what python provides and I get the rationale but Isaacs comments left me with the impression that node has a way more centralized file system namespace than what python has: https://docs.python.org/3.7/library/filesys.html 

so in my understanding it seems to be more friendly to node developers to implement (all the `shutil` family of functions) in `fs` as suggested, is that impression something any of you share?